packages/hardhat-waffle/test/fixture-projects/hardhat-project/test/test-link.js

```diff
@@ -0,0 +1,31 @@
+const { expect, use } = require('chai');
+const { waffle, ethers, artifacts } = require('hardhat');
+
+const { BigNumber } = ethers;
+const { solidity, deployContract, getSigner, link, provider } = waffle;
+
+use(solidity);
+
+describe('test contract linking', async () => {
+  const signer = provider.getSigner();
+
+  it('deployContract without libraries works', async () => {
+    const contract = await deployContract(signer, await artifacts.readArtifact('Contract'));
+    await expect(contract.inc(7)).to.emit(contract, 'Increment').withArgs(7);
+  });
+
+  it('should be able to link libraries', async () => {
+    const lib = await deployContract(signer, await artifacts.readArtifact('Lib'));
+    expect(lib.address).to.be.properAddress;
+
+    // no straightforward way to use hardhat-waffle's `link` library from here
+    // with deployContract or readArtifact
+    // link(UsesLibArtifact, 'contract/path', lib.address);
```

Comment:
Hey, thanks for sending this PR, @clifton. I love the idea of a failing test as a report/feature request 😍

Can you link to waffles config about this or explain a little more? I'm not sure I understand this.

Reply:
Sure. External libraries from npm packages work fine with deployContract but if you try to include a library from a relative path, it throws an invalid bytecode error on deployment.

I understand that ether’s getContractFactory will work with the link function here, though it’s not clear in the docs or in hardhat’s API what the idiomatic way of instantiating contracts for testing is. If it’s with getContractFactory, why does the deployContract function in hardhat-waffle exist? Or should all linking happen automatically?

Reply:
I believe this is a problem in Waffle. I opened an issue on their repo: https://github.com/EthWorks/Waffle/issues/429

Reply:
Yep, makes total sense. Thank you!